File: 08. Trees/02. Inorder Traversal/main.js
class TreeNode {
    constructor(val) {
        this.val = val;
        this.left = this.right = null;
    }
}

// Inorder Traversal (Recursive)
function inorderTraversalRecursive(root, result) {
    if (!root) return;
    inorderTraversalRecursive(root.left, result);
    result.push(root.val);
    inorderTraversalRecursive(root.right, result);
}

// Inorder Traversal (Iterative)
function inorderTraversalIterative(root) {
    let result = [];
    let stack = [];
    let current = root;

    while (current !== null || stack.length > 0) {
        while (current !== null) {
            stack.push(current);
            current = current.left;
        }
        current = stack.pop();
        result.push(current.val);
        current = current.right;
    }
    return result;
}

// Main function
function main() {
    let root = new TreeNode(1);
    root.left = new TreeNode(2);
    root.right = new TreeNode(3);
    root.left.left = new TreeNode(4);
    root.left.right = new TreeNode(5);
    root.right.left = new TreeNode(6);
    root.right.right = new TreeNode(7);

    let recursiveResult = [];
    inorderTraversalRecursive(root, recursiveResult);
    console.log("Recursive Inorder Traversal:", recursiveResult);

    let iterativeResult = inorderTraversalIterative(root);
    console.log("Iterative Inorder Traversal:", iterativeResult);
}

// Run the main function
main();

File: 08. Trees/02. Inorder Traversal/main.cpp
#include <bits/stdc++.h>
using namespace std;

// Definition of TreeNode
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

// Inorder Traversal (Recursive)
void inorderTraversalRecursive(TreeNode* root, vector<int>& result) {
    if (!root) return;
    inorderTraversalRecursive(root->left, result);
    result.push_back(root->val);
    inorderTraversalRecursive(root->right, result);
}

// Inorder Traversal (Iterative)
vector<int> inorderTraversalIterative(TreeNode* root) {
    vector<int> result;
    stack<TreeNode*> st;
    TreeNode* current = root;

    while (current || !st.empty()) {
        while (current) {  // Reach the leftmost node
            st.push(current);
            current = current->left;
        }
        current = st.top();
        st.pop();
        result.push_back(current->val);
        current = current->right; // Move to right subtree
    }
    return result;
}

int main() {
    // Sample tree:   1
    //              /   \
    //             2     3
    //            / \   / \
    //           4   5 6   7

    TreeNode* root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);
    root->left->left = new TreeNode(4);
    root->left->right = new TreeNode(5);
    root->right->left = new TreeNode(6);
    root->right->right = new TreeNode(7);

    vector<int> recursiveResult;
    inorderTraversalRecursive(root, recursiveResult);
    cout << "Recursive Inorder Traversal: ";
    for (int val : recursiveResult) cout << val << " ";
    cout << endl;

    vector<int> iterativeResult = inorderTraversalIterative(root);
    cout << "Iterative Inorder Traversal: ";
    for (int val : iterativeResult) cout << val << " ";
    cout << endl;

    return 0;
}
